File: pst/actions/transfer.ts
import {ActionSchema, HandlerResultSchema, StateSchema} from "../schemas";
import {Transaction} from "../imports/smartweave/transaction";

export function transfer(state: StateSchema, action: ActionSchema): HandlerResultSchema {
    const target = action.target;
    const qty = action.qty;
    const caller = Transaction.owner();

    if (qty <= 0 || caller === target) {
        throw new Error('[CE:ITT] Invalid token transfer');
    }

    if (state.balances.get(caller) < qty) {
        throw new Error(`[CE:NEB] Caller balance not high enough to send ${qty} token(s)!`);
    }

    // Lower the token balance of the caller
    state.balances.set(caller, state.balances.get(caller) - qty);
    if (!state.balances.has(target)) {
        state.balances.set(target,qty);
    } else {
        state.balances.set(target, state.balances.get(target) + qty);
    }

    return {
        state,
        result: null
    };
}

File: rust/src/actions/transfer.rs
use crate::error::ContractError;
use crate::error::ContractError::{CallerBalanceNotEnough, TransferAmountMustBeHigherThanZero};
use crate::state::{State, HandlerResult};
use crate::contract_utils::js_imports::{log, Transaction};

pub fn transfer(mut state: State, qty: u64, target: String) -> Result<HandlerResult, ContractError> {
    log(&format!("Transfer called: {}: {}", target, qty));

    if qty == 0 {
        return Err(TransferAmountMustBeHigherThanZero);
    }

    let caller = Transaction::owner();
    let balances = &mut state.balances;

    // Checking if caller has enough funds
    let caller_balance = *balances.get(&caller).unwrap_or(&0);
    if caller_balance < qty {
        return Err(CallerBalanceNotEnough(caller_balance));
    }

    // Update caller balance
    balances.insert(caller, caller_balance - qty);

    // Update target balance
    let target_balance = *balances.get(&target).unwrap_or(&0);
    balances.insert(target, target_balance + qty);

    // Debug logging
    for (key, value) in &state.balances {
        log(&format!("{}: {}", key, value));
    }

    Ok(HandlerResult::NewState(state))
}
